### ISSUE TITLE
[BUG] All Browserless invocations throw 401 error in v2 workflows

### ISSUE THREAD

**Issue by jeffdechambeau** [PipedreamHQ/pipedream#2983]
**Description**
v2 workflows throw 401 errors for all Browserless calls. 

**To Reproduce**

1. Go to https://browserless.io and get an api key
2. Create a v2 workflow
3. Search for Browserless
4. Choose/select any preset
5. Populate with prop for API key
6. Hit test 
7. Experience bug

**Expected behavior**
That's probably beyond the scope of this ticket, suffice it to say the function call should do thing instead of throwing a 401 error. 

**Error Log**
```
ErrorEvent {
  [Symbol(kTarget)]: WebSocket {
    _events: [Object: null prototype] { open: [Function], error: [Function] },
    _eventsCount: 2,
    _maxListeners: undefined,
    _binaryType: 'nodebuffer',
    _closeCode: 1006,
    _closeFrameReceived: false,
    _closeFrameSent: false,
    _closeMessage: <Buffer >,
    _closeTimer: null,
    _extensions: {},
    _paused: false,
    _protocol: '',
    _readyState: 3,
    _receiver: null,
    _sender: null,
    _socket: null,
    _bufferedAmount: 0,
    _isServer: false,
    _redirects: 0,
    _url: 'wss://chrome.browserless.io?token=########-####-####-####-#########', // appears in full & correctly in original error
    _originalHost: 'chrome.browserless.io',
    _req: ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 4,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: true,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: true,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: false,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: 0,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      socket: [TLSSocket],
      _header: 'GET /?token=########-####-####-####-######### HTTP/1.1\r\n' +
        'Sec-WebSocket-Version: 13\r\n' +
        'Sec-WebSocket-Key: rlItjKFfnNgRCT2T7JgJQQ==\r\n' +
        'Connection: Upgrade\r\n' +
        'Upgrade: websocket\r\n' +
        'User-Agent: Puppeteer 14.1.1\r\n' +
        'Host: chrome.browserless.io\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: noopPendingOutput],
      agent: undefined,
      socketPath: undefined,
      method: 'GET',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      path: '/?token=########-####-####-####-#########',
      _ended: true,
      res: [IncomingMessage],
      aborted: true,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: false,
      host: 'chrome.browserless.io',
      protocol: 'https:',
      [Symbol(kCapture)]: false,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype],
      [Symbol(kAborted)]: true
    },
    [Symbol(kCapture)]: false
  },
  [Symbol(kType)]: 'error',
  [Symbol(kError)]:     at ClientRequest.null (/tmp/ee/c_jGfanAPy/node_modules/.pnpm/ws@8.6.0/node_modules/ws/lib/websocket.js:873:7)
      at ClientRequest.emit (events.js:400:28)
      at HTTPParser.parserOnIncomingClient (_http_client.js:647:27)
      at HTTPParser.parserOnHeadersComplete (_http_common.js:127:17)
      at TLSSocket.socketOnData (_http_client.js:515:22)
      at TLSSocket.emit (events.js:400:28)
      at null.addChunk (internal/streams/readable.js:293:12)
      at null.readableAddChunk (internal/streams/readable.js:267:9)
      at TLSSocket.Readable.push (internal/streams/readable.js:206:10)
      at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)
  ,
  [Symbol(kMessage)]: 'Unexpected server response: 401'
}

```

**Desktop:**
 - OS: MacOS
 - Browser: Chrome
 - Version 101.0.4951.64 (Official Build) (x86_64)

**Additional context**
- Errors when using default component configuration
- Also errors after directly copying over the import + call from a v1 workflow (& updating the api_key ref)
- Also fails running it as a normal node step w/ import
- API key(s) were tested in v1 workflows and are valid

**Comment by jeffdechambeau** [comment]
Thanks @dannyroosevelt -- that's the step, yes. 

I just tried again but am still having the same error. 

I deleted and re-added the steps to check, but no dice. 

please lmk if I can provide any additional useful context.

**Comment by dannyroosevelt** [comment]
Hm that's strange. This is what I see:
![image](https://user-images.githubusercontent.com/20780753/170800734-bfbaff50-1496-412f-98aa-bd00c35c0453.png)

Is this the code you're trying?
```
module.exports = defineComponent({
  props: {
    browserless: {
      type: "app",
      app: "browserless",
    }
  },
  async run({steps, $}) {
    // See the browserless docs for more info:
    // https://docs.browserless.io/
    const puppeteer = require('puppeteer-core') 
    
    const browser = await puppeteer.connect({ 
      browserWSEndpoint: `wss://chrome.browserless.io?token=${this.browserless.$auth.api_key}` 
    })
    const page = await browser.newPage()
    
    const url = "https://example.com"
    const type = "png"
  
    await page.goto(url)
    const screenshot = await page.screenshot()
    
    // export the base64-encoded screenshot for use in future steps,
    // along with the image type and filename
    $.export("screenshot", Buffer.from(screenshot, 'binary').toString('base64'))
    $.export("type", type)
    $.export("filename",`${url.replace(/[&\/\\#, +()$~%.'":*?<>{}]/g, '_')}-${+new Date()}.${type}`)
    
    await browser.close()
  },
})
```

**Comment by jeffdechambeau** [comment]
Here's a screen recording of me creating a new workflow and adding the default browserless step, configuring it and hitting test: 

Here's the code so you don't have to freezferame:

`    // See the browserless docs for more info:
    // https://docs.browserless.io/
    const puppeteer = require('puppeteer-core') 
    
    const browser = await puppeteer.connect({ 
      browserWSEndpoint: `wss://chrome.browserless.io?token=${this.browserless.$auth.api_key}` 
    })
    const page = await browser.newPage()
    
    const url = "https://example.com"
    const type = "png"
  
    await page.goto(url)
    const screenshot = await page.screenshot()
    
    // export the base64-encoded screenshot for use in future steps,
    // along with the image type and filename
    $.export("screenshot", Buffer.from(screenshot, 'binary').toString('base64'))
    $.export("type", type)
    $.export("filename",`${url.replace(/[&\/\\#, +()$~%.'":*?<>{}]/g, '_')}-${+new Date()}.${type}`)
    
    await browser.close()`

Looks the same to my eye.

**Comment by jeffdechambeau** [comment]
Sorry, this is my first ever github ticket.

**Comment by ctrlaltdylan** [comment]
HI @jeffdechambeau ,

Just to add some color, unless otherwise specified in the import, Pipedream will automatically download the latest version of the dependencies listed in the require/import statements.

I was able to use browserless by replacing the `puppeeter-core` dependency with the normal `puppeteer` dependency:

```javascript
   
import puppeteer from 'puppeteer'

export default defineComponent({
  props: {
    browserless: {
      type: "app",
      app: "browserless",
    }
  },
  async run({steps, $}) {
    // See the browserless docs for more info:
    // https://docs.browserless.io/
    
    const browser = await puppeteer.connect({ 
      browserWSEndpoint: `wss://chrome.browserless.io?token=${this.browserless.$auth.api_key}` 
    })
    const page = await browser.newPage()
    
    const url = "https://google.com"
    const type = "png"
  
    await page.goto(url)
    const screenshot = await page.screenshot()
    
    // export the base64-encoded screenshot for use in future steps,
    // along with the image type and filename
    $.export("screenshot", Buffer.from(screenshot, 'binary').toString('base64'))
    $.export("type", type)
    $.export("filename",`${url.replace(/[&\/\\#, +()$~%.'":*?<>{}]/g, '_')}-${+new Date()}.${type}`)
    
    await browser.close()
  },
})
```

Could you replace your code with this snippet above and let us know if that fixes it for you?

**Comment by jeffdechambeau** [comment]
Hey Dylan @ctrlaltdylan thank you for the snippet. Unfortunately no dice: 

<img width="445" alt="Screen Shot 2022-05-31 at 11 45 05 AM" src="https://user-images.githubusercontent.com/1578544/171214853-9233a9f3-de47-4596-b19e-0e0f0ca9ab82.png">

**Comment by ctrlaltdylan** [comment]
Thanks for sharing your results @jeffdechambeau but unfortunately I'm not able to reproduce with the same code.

It's silly, but could you double check there aren't any newlines or extra / missing characters from your Browserless.io API key in your **Connected Accounts** in your Pipedream dashboard?

![Uploading CleanShot 2022-05-31 at 11.52.08@2x.png…]()

Also, it might be worth logging into the Browserless API dashboard and checking to make sure you haven't hit an API usage limit there.

**Comment by jeffdechambeau** [comment]
@dylburger I was able to get it working, by not using any prop functionality.

I believe the issue may be with the props injection logic: if I define the apiKey locally in the function and then insert it into the request string, it works. Similarly if I hardcore the entire API request url, it works. When I add back the prop ref, it fails. 

I also checked for hidden whitespace but but strings are exactly 36ch, nothing hidden or unexpected. I tested this across both keys, and confirmed that both keys work as of right now. They do. 

When I consoled logged the prop instance of the apiKey and copied it to check from whitespace, I found that some had been appended, but idk if that was from the logging UI, or an earlier step. 

Here's what I wound up doing to make it work: 

<img width="606" alt="Screen Shot 2022-05-31 at 12 25 04 PM" src="https://user-images.githubusercontent.com/1578544/171225052-7beb023a-24c1-4e9d-ad68-d0dd1d55fb6b.png">

I don't know what protocol is in terms of closing this ticket, but I'm good! I'm available to provide more info if it's useful for your end either here on on Slack under the same username. 

Pipedream has been a gamechanger and I have a growing cohort of clients now using and loving it, thanks so much for making (and attentively supporting) such a fantastic service!
